pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,180 @@
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	client    client.Client
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.client.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// List all the Namespaces which matches the Namespace LabelSelector.
+	namespaceList := corev1.NamespaceList{}
+	if ingressController.Spec.NamespaceSelector != nil {
+		namespaceLabelSelector := ingressController.Spec.NamespaceSelector.MatchLabels
+		if err := r.client.List(ctx, &namespaceList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(namespaceLabelSelector),
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %v", request, err)
+		}
+	} else {
+		// If Namespace LabelSelector is not provided then add Namespace with empty name to indicate all Namespaces.
+		namespaceList.Items = append(namespaceList.Items, corev1.Namespace{})
+	}
+
+	// Get the Route LabelSelector from the Ingress Controller.
+	var routeLabelSelector map[string]string
+	if ingressController.Spec.RouteSelector != nil {
+		routeLabelSelector = ingressController.Spec.RouteSelector.MatchLabels
+	}
+
+	// Store the previous value of the metric.
+	prevMetricValue := GetRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Initialize the RouteMetricsControllerRoutesPerShard metric for Ingress Controller.
+	InitializeRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Iterate through all the Namespaces.
+	for _, namespace := range namespaceList.Items {
+		// List all the Routes which matches the Route LabelSelector and the Namespace.
+		routeList := routev1.RouteList{}
+		if err := r.client.List(ctx, &routeList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(routeLabelSelector),
+			Namespace:     namespace.Name,
+		}); err != nil {
+			// Set the value of the metric to its previous value and re-queue the request.
+			SetRouteMetricsControllerRoutesPerShardMetric(request.Name, prevMetricValue)
```

Comment:
Do you have to write this value again?  What happens if you don't, won't the old value be there anyway?

Reply:
I'll change the implementation to using a counter variable to keep track of the number of routes and setting the value of the metric after the completion of the loop. Then this line would not be required.

Reply:
The counter variable is not preferred for this because it can only go up.  We use gauge for metrics that can go up and down.

Reply:
I meant a temporary Go variable for counting the routes and then setting the metric value at the end of the loop. The metric type would still be `Gauge` and not `Counter`.